### PR TITLE
Assert conn_pool_stream_handle_ is nullptr

### PR DESCRIPTION
**Pull request by rgs1**
Ensures we never reach an invalid state. 

Somewhat related to #13737. 

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
(Possibly not ready for review yet, need to ensure integ is happy).

**Comment by rgs1** [comment]
OS X timed out building:

https://dev.azure.com/cncf/envoy/_build/results?buildId=55383&view=logs&j=fa3d3e18-6969-5713-c3e7-3581195704fd&t=ba370bef-ca49-52f5-0c93-7c0c0f27c465

File: test/common/router/router_test.cc [hunk 1]
```diff
@@ -641,7 +641,7 @@ TEST_F(RouterTest, AddCookie) {
               Http::ConnectionPool::Callbacks& callbacks) -> Http::ConnectionPool::Cancellable* {
             response_decoder = &decoder;
             callbacks.onPoolReady(encoder, cm_.conn_pool_.host_, upstream_stream_info_);
-            return &cancellable_;
+            return nullptr;
```
Comment (rgs1) [inline test/common/router/router_test.cc:644]: If I return the cancellable_ here, the assertion fails. I am guessing when the cancellable is returned, something is calling newStream() on it without a subsequent onPoolReady() or onPoolFailure().

@alyssawilk is this correct in the context of this test?

Comment (alyssawilk) [reply test/common/router/router_test.cc:644]: Unsure of the details but I'm pretty sure the only thing that really matters in this test is the Max-Age=1337; etc. cookie logic below :-)

File: source/extensions/upstreams/http/http/upstream_request.h [hunk 2]
```diff
@@ -26,6 +27,9 @@ class HttpConnPool : public Router::GenericConnPool, public Envoy::Http::Connect
     conn_pool_ = cm.httpConnPoolForCluster(route_entry.clusterName(), route_entry.priority(),
                                            downstream_protocol, ctx);
   }
+  ~HttpConnPool() override {
+    RELEASE_ASSERT(conn_pool_stream_handle_ == nullptr, "conn_pool_stream_handle not null");
```
Comment (alyssawilk) [inline source/extensions/upstreams/http/http/upstream_request.h:31]: Preference for ASSERT, or you can use ENVOY_BUG if you think it's tracking-worthy.

**Review by alyssawilk — COMMENTED** [review]
Looks good - I'm always a fan of more assertion checking - just gives us more things the integrations tests can find :-)

File: source/extensions/upstreams/http/http/upstream_request.h (hunk 2 above)
Comment (rgs1) [reply source/extensions/upstreams/http/http/upstream_request.h]: Fair, updated.

File: test/common/router/router_test.cc (hunk 1 above)
Comment (rgs1) [reply test/common/router/router_test.cc:644]: Ah ok, great.

**Review by alyssawilk — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**